project-code/app/com/edulify/modules/geolocation/GeolocationService.java

```diff
@@ -1,14 +1,27 @@
 package com.edulify.modules.geolocation;
 
-/**
- * @deprecated Use AsyncGeolocationService instead
- */
-@Deprecated
-public class GeolocationService {
+import play.libs.F.Promise;
 
-  private static final long timeout = Config.getMillisecondsOr("geolocation.timeout", 5000);
+import static play.libs.F.Promise.pure;
 
-  public static Geolocation getGeolocation(String ip) {
-    return AsyncGeolocationService.getGeolocation(ip).get(timeout);
+public final class GeolocationService {
+
+  private final GeolocationCache cache;
+  private final GeolocationProvider provider;
+
+  public GeolocationService(GeolocationCache cache, GeolocationProvider provider) {
+    this.cache = cache;
+    this.provider = provider;
+  }
+
+  public Promise<Geolocation> getGeolocation(String ip) {
+    if (null == ip) return pure(null);
+    if (ip.isEmpty()) return pure(null);
+    Geolocation geolocation = cache.get(ip);
```

Comment:
Maybe `GeolocationCache` should expose a `getOrElse` method, like Play cache does, and then do something like:

```java
cache.getOrElse(ip, ipToGeolocate -> { ... });
```